src/daft-sql/src/planner.rs

```diff
@@ -694,48 +739,47 @@ impl SQLPlanner {
+                        (
                             sqlparser::ast::Expr::Identifier(left),
                             sqlparser::ast::Expr::Identifier(right),
-                        ) = (left.as_ref(), right.as_ref())
-                        {
-                            let left = ident_to_str(left);
-                            let right = ident_to_str(right);
-
-                            // we don't know which table the identifiers belong to, so we need to check both
-                            let left_schema = left_rel.schema();
-                            let right_schema = right_rel.schema();
-
-                            // if the left side is in the left schema, then we assume the right side is in the right schema
-                            let (left_on, right_on) = if left_schema.get_field(&left).is_ok() {
-                                (col(left), col(right))
-                            // if the right side is in the left schema, then we assume the left side is in the right schema
-                            } else if right_schema.get_field(&left).is_ok() {
-                                (col(right), col(left))
-                            } else {
-                                unsupported_sql_err!("JOIN clauses must reference columns in the joined tables; found `{}`", left);
-                            };
+                        ) =>{
+                            collect_idents(&[left.clone()], &[right.clone()], left_rel, right_rel)
+                                .map(|(left, right)| (left, right, vec![null_equals_null]))
 
-                            let null_equals_null = *op == BinaryOperator::Spaceship;
+                        }
+                        (
+                            sqlparser::ast::Expr::CompoundIdentifier(left),
+                            sqlparser::ast::Expr::Identifier(right)
+                        ) => {
+                            collect_idents(left, &[right.clone()], left_rel, right_rel)
+                                .map(|(left, right)| (left, right, vec![null_equals_null]))
 
-                            Ok((vec![left_on], vec![right_on], vec![null_equals_null]))
-                        } else {
-                            unsupported_sql_err!("JOIN clauses support '='/'<=>' constraints on identifiers; found `{left} {op} {right}`");
+                        }
+                        (
+                            sqlparser::ast::Expr::Identifier(left),
+                            sqlparser::ast::Expr::CompoundIdentifier(right)
+                        ) => {
+                            collect_idents(&[left.clone()], right, left_rel, right_rel)
+                                .map(|(left, right)| (left, right, vec![null_equals_null]))
+                        }
+                        _ => unsupported_sql_err!("process_join_on: Expected CompoundIdentifier, but found left: {:?}, right: {:?}", left, right),
                         }
```

Comment:
To avoid doing this large match, I'm wondering if we could do something like this instead
```rust
fn get_idents_vec(expr: sqlparser::ast::Expr) -> &[Ident] {
    match expr {
        sqlparser::ast::Expr::Identifier(ident) => &[ident],
        sqlparser::ast::Expr::CompoundIdentifier(idents) => idents,
        _ => unsupported_sql_err!()
    }
}

let left = get_idents_vec(left);
let right = get_idents_vec(right);

collect_idents(left, right, left_rel, right_rel)
```

Reply:
thats a good idea. I didn't like this match statement either. Will update!